File: scheduler.py
"""
CPU Scheduling Simulator: Round Robin vs Shortest Job First (Non-Preemptive)
OS Course Project — Full Submission
"""

def get_valid_int(prompt, min_val=None, allow_zero=False):
    while True:
        raw = input(prompt).strip()
        if not raw:
            print("  [Error] Input cannot be empty.")
            continue
        try:
            val = int(raw)
        except ValueError:
            print("  [Error] Must be a whole number (no decimals, no letters).")
            continue
        if not allow_zero and min_val is not None and val < min_val:
            print(f"  [Error] Value must be >= {min_val}.")
            continue
        if allow_zero and min_val is not None and val < min_val:
            print(f"  [Error] Value must be >= {min_val}.")
            continue
        return val


def input_processes():
    print("\n" + "="*60)
    print("  PROCESS INPUT")
    print("="*60)
    n = get_valid_int("Enter number of processes (1-10): ", min_val=1)
    if n > 10:
        print("  [Error] Maximum 10 processes allowed.")
        return input_processes()

    processes = []
    pids = set()

    for i in range(n):
        print(f"\n  --- Process {i+1} ---")
        while True:
            pid = input("  Process ID (e.g. P1): ").strip()
            if not pid:
                print("  [Error] Process ID cannot be empty.")
                continue
            if pid in pids:
                print(f"  [Error] Duplicate Process ID '{pid}'. Use a unique ID.")
                continue
            pids.add(pid)
            break

        at = get_valid_int("  Arrival Time (>= 0): ", min_val=0, allow_zero=True)
        bt = get_valid_int("  Burst Time (>= 1): ", min_val=1)
        processes.append({"pid": pid, "at": at, "bt": bt})

    return processes


def input_quantum():
    print()
    return get_valid_int("Enter Time Quantum for Round Robin (>= 1): ", min_val=1)


# ─── Round Robin ────────────────────────────────────────────────────────────

def run_round_robin(processes, quantum):
    n = len(processes)
    remaining = {p["pid"]: p["bt"] for p in processes}
    first_run = {}
    finish = {}
    gantt = []
    t = 0
    queue = []
    in_queue = set()
    arrived = set()

    procs_sorted = sorted(processes, key=lambda p: p["at"])

    def enqueue_arrivals(time):
        for p in procs_sorted:
            if p["pid"] not in arrived and p["at"] <= time and p["pid"] not in in_queue and remaining[p["pid"]] > 0:
                queue.append(p["pid"])
                in_queue.add(p["pid"])
                arrived.add(p["pid"])

    enqueue_arrivals(t)
    if not queue:
        t = procs_sorted[0]["at"]
        enqueue_arrivals(t)

    iterations = 0
    while queue and iterations < 100000:
        iterations += 1
        pid = queue.pop(0)
        in_queue.discard(pid)
        p = next(x for x in processes if x["pid"] == pid)

        if pid not in first_run:
            first_run[pid] = t

        run = min(remaining[pid], quantum)
        gantt.append((pid, t, t + run))
        t += run
        remaining[pid] -= run

        enqueue_arrivals(t)

        if remaining[pid] > 0:
            queue.append(pid)
            in_queue.add(pid)
        else:
            finish[pid] = t

        if not queue and any(remaining[p["pid"]] > 0 for p in processes):
            unfinished = [p for p in processes if remaining[p["pid"]] > 0 and p["pid"] not in in_queue]
            if unfinished:
                next_at = min(p["at"] for p in unfinished)
                if next_at > t:
                    t = next_at
                enqueue_arrivals(t)

    results = []
    for p in processes:
        pid = p["pid"]
        wt = finish[pid] - p["at"] - p["bt"]
        tat = finish[pid] - p["at"]
        rt = first_run[pid] - p["at"]
        results.append({"pid": pid, "at": p["at"], "bt": p["bt"], "wt": wt, "tat": tat, "rt": rt, "ft": finish[pid]})
    return results, gantt


# ─── SJF Non-Preemptive ─────────────────────────────────────────────────────

def run_sjf(processes):
    n = len(processes)
    done = set()
    results = []
    gantt = []
    t = 0

    for _ in range(n):
        available = [p for p in processes if p["at"] <= t and p["pid"] not in done]
        if not available:
            next_p = min((p for p in processes if p["pid"] not in done), key=lambda x: x["at"])
            t = next_p["at"]
            available = [next_p]

        # Shortest burst; tie-break by arrival time then pid
        chosen = min(available, key=lambda p: (p["bt"], p["at"], p["pid"]))
        start = t
        t += chosen["bt"]
        gantt.append((chosen["pid"], start, t))
        wt = start - chosen["at"]
        tat = t - chosen["at"]
        rt = start - chosen["at"]  # non-preemptive: RT == WT
        results.append({"pid": chosen["pid"], "at": chosen["at"], "bt": chosen["bt"],
                        "wt": wt, "tat": tat, "rt": rt, "ft": t})
        done.add(chosen["pid"])

    return results, gantt


# ─── Display ─────────────────────────────────────────────────────────────────

def print_gantt(gantt, title):
    print(f"\n  Gantt Chart — {title}")
    bar = ""
    timeline = ""
    for pid, start, end in gantt:
        width = max(len(pid) + 2, (end - start) * 3)
        bar += f"|{pid.center(width)}"
        timeline += str(start).ljust(width + 1)
    bar += "|"
    timeline += str(gantt[-1][2])
    print("  " + bar)
    print("  " + timeline)


def print_metrics(results, title):
    print(f"\n  Metrics — {title}")
    header = f"  {'PID':<8} {'AT':>5} {'BT':>5} {'FT':>5} {'WT':>6} {'TAT':>6} {'RT':>6}"
    print(header)
    print("  " + "-" * 48)
    for r in results:
        print(f"  {r['pid']:<8} {r['at']:>5} {r['bt']:>5} {r['ft']:>5} {r['wt']:>6} {r['tat']:>6} {r['rt']:>6}")
    print("  " + "-" * 48)
    avg_wt  = sum(r["wt"]  for r in results) / len(results)
    avg_tat = sum(r["tat"] for r in results) / len(results)
    avg_rt  = sum(r["rt"]  for r in results) / len(results)
    print(f"  {'Average':<8} {'':>5} {'':>5} {'':>5} {avg_wt:>6.2f} {avg_tat:>6.2f} {avg_rt:>6.2f}")
    return avg_wt, avg_tat, avg_rt


def print_comparison(rr_avgs, sjf_avgs):
    print("\n" + "="*60)
    print("  COMPARISON SUMMARY — RR vs SJF (same workload)")
    print("="*60)
    labels = ["Avg Waiting Time", "Avg Turnaround Time", "Avg Response Time"]
    rr_vals  = [rr_avgs[0],  rr_avgs[1],  rr_avgs[2]]
    sjf_vals = [sjf_avgs[0], sjf_avgs[1], sjf_avgs[2]]

    print(f"\n  {'Metric':<22} {'Round Robin':>12} {'SJF':>12} {'Winner':>10}")
    print("  " + "-"*58)
    for label, rr, sjf in zip(labels, rr_vals, sjf_vals):
        if rr < sjf:
            winner = "RR"
        elif sjf < rr:
            winner = "SJF"
        else:
            winner = "Tie"
        print(f"  {label:<22} {rr:>12.2f} {sjf:>12.2f} {winner:>10}")

    print("\n  SCHEDULING ANALYSIS:")
    print("  - SJF minimizes average waiting time when all jobs are known upfront.")
    print("  - Round Robin ensures fairness — no process waits indefinitely (no starvation).")
    print("  - SJF can cause starvation for long processes if short jobs keep arriving.")
    print("  - Round Robin's performance depends heavily on the chosen time quantum.")
    print("  - RR has higher overhead due to frequent context switches.")
    print("  - SJF is optimal for batch systems; RR is preferred for time-sharing systems.")


# ─── Main ────────────────────────────────────────────────────────────────────

def main():
    print("\n" + "="*60)
    print("  CPU SCHEDULING SIMULATOR")
    print("  Round Robin vs Shortest Job First (Non-Preemptive)")
    print("="*60)

    processes = input_processes()
    quantum = input_quantum()

    print("\n" + "="*60)
    print("  ROUND ROBIN RESULTS")
    print("="*60)
    rr_results, rr_gantt = run_round_robin(processes, quantum)
    print_gantt(rr_gantt, f"Round Robin (Quantum = {quantum})")
    rr_avgs = print_metrics(rr_results, "Round Robin")

    print("\n" + "="*60)
    print("  SJF (NON-PREEMPTIVE) RESULTS")
    print("="*60)
    sjf_results, sjf_gantt = run_sjf(processes)
    print_gantt(sjf_gantt, "SJF Non-Preemptive")
    sjf_avgs = print_metrics(sjf_results, "SJF Non-Preemptive")

    print_comparison(rr_avgs, sjf_avgs)
    print("\n" + "="*60 + "\n")


if __name__ == "__main__":
    main()

File: Web/src/scheduler.js
export function runRoundRobin(processes, quantum) {
    const n = processes.length;
    const remaining = {};
    processes.forEach(p => remaining[p.pid] = p.bt);
    
    const firstRun = {};
    const finish = {};
    const gantt = [];
    const trace = [];
    
    const queue = [];
    const inQueue = new Set();
    const arrived = new Set();
    let t = 0;
    
    // Sort processes by arrival time, then PID (to match Python behavior)
    const procsByAt = [...processes].sort((a, b) => {
        if (a.at !== b.at) return a.at - b.at;
        return a.pid.localeCompare(b.pid);
    });

    function enqueueArrivals(time) {
        for (const p of procsByAt) {
            if (!arrived.has(p.pid) && p.at <= time && !inQueue.has(p.pid) && remaining[p.pid] > 0) {
                queue.push(p.pid);
                inQueue.add(p.pid);
                arrived.add(p.pid);
            }
        }
    }

    enqueueArrivals(t);

    if (queue.length === 0) {
        if (procsByAt.length > 0) {
            t = procsByAt[0].at;
            enqueueArrivals(t);
        }
    }

    let guard = 0;
    while (queue.length > 0 && guard < 100000) {
        guard++;
        const pid = queue.shift();
        inQueue.delete(pid);
        
        const p = processes.find(x => x.pid === pid);
        
        // Record trace BEFORE execution
        trace.push({
            time: t,
            running: pid,
            queueSnapshot: [...queue]
        });

        if (!(pid in firstRun)) {
            firstRun[pid] = t;
        }

        const run = Math.min(remaining[pid], quantum);
        gantt.push({ pid, start: t, end: t + run });
        
        t += run;
        remaining[pid] -= run;

        enqueueArrivals(t);

        if (remaining[pid] > 0) {
            queue.push(pid);
            inQueue.add(pid);
        } else {
            finish[pid] = t;
        }

        if (queue.length === 0) {
            const pending = processes.filter(x => remaining[x.pid] > 0 && !inQueue.has(x.pid));
            if (pending.length > 0) {
                const nextAt = Math.min(...pending.map(x => x.at));
                if (nextAt > t) {
                    t = nextAt;
                }
                enqueueArrivals(t);
            }
        }
    }

    const results = processes.map(p => {
        const ft = finish[p.pid];
        const wt = ft - p.at - p.bt;
        const tat = ft - p.at;
        const rt = firstRun[p.pid] - p.at;
        return {
            pid: p.pid,
            at: p.at,
            bt: p.bt,
            ft,
            wt,
            tat,
            rt
        };
    });

    return { results, gantt, trace };
}

export function runSJF(processes) {
    const done = new Set();
    const results = [];
    const gantt = [];
    let t = 0;

    for (let i = 0; i < processes.length; i++) {
        let available = processes.filter(p => p.at <= t && !done.has(p.pid));

        if (available.length === 0) {
            const remainingProcs = processes.filter(p => !done.has(p.pid));
            if (remainingProcs.length === 0) break;
            
            // Find next arrival time
            let nextP = remainingProcs[0];
            for (const p of remainingProcs) {
                if (p.at < nextP.at || (p.at === nextP.at && p.pid.localeCompare(nextP.pid) < 0)) {
                    nextP = p;
                }
            }
            t = nextP.at;
            available = processes.filter(p => p.at <= t && !done.has(p.pid));
        }

        // Tie-breaking: BT -> AT -> PID
        let chosen = available[0];
        for (const p of available) {
            if (p.bt < chosen.bt) {
                chosen = p;
            } else if (p.bt === chosen.bt) {
                if (p.at < chosen.at) {
                    chosen = p;
                } else if (p.at === chosen.at) {
                    if (p.pid.localeCompare(chosen.pid) < 0) {
                        chosen = p;
                    }
                }
            }
        }

        const start = t;
        t += chosen.bt;
        gantt.push({ pid: chosen.pid, start, end: t });

        const wt = start - chosen.at;
        const tat = t - chosen.at;
        const rt = start - chosen.at;

        results.push({
            pid: chosen.pid,
            at: chosen.at,
            bt: chosen.bt,
            ft: t,
            wt,
            tat,
            rt
        });
        done.add(chosen.pid);
    }

    return { results, gantt };
}

export function calculateAverages(results) {
    if (!results || results.length === 0) return { wt: 0, tat: 0, rt: 0 };
    const n = results.length;
    const avgWt = results.reduce((acc, curr) => acc + curr.wt, 0) / n;
    const avgTat = results.reduce((acc, curr) => acc + curr.tat, 0) / n;
    const avgRt = results.reduce((acc, curr) => acc + curr.rt, 0) / n;
    return { wt: avgWt, tat: avgTat, rt: avgRt };
}
